File: course_4/module_3/examples/greedy_knapsack_heuristic.go
package main

import (
	"fmt"
	"sort"
)

// Item represents an item in the knapsack problem.
type Item struct {
	Value  int     // Value of the item.
	Weight int     // Weight of the item.
	Ratio  float64 // Value-to-weight ratio of the item.
}

// Knapsack represents the knapsack used to store selected items.
type Knapsack struct {
	Capacity      int    // Total capacity of the knapsack.
	CurrentWeight int    // Current weight of the knapsack.
	Items         []Item // List of items included in the knapsack.
	TotalValue    int    // Total value of the items in the knapsack.
}

// NewKnapsack initializes a new Knapsack.
func NewKnapsack(capacity int) *Knapsack {
	if capacity <= 0 {
		panic("Knapsack capacity must be positive.")
	}
	return &Knapsack{
		Capacity: capacity,
	}
}

// AddItem tries to add an item to the knapsack if it fits.
func (k *Knapsack) AddItem(item Item) bool {
	if k.CurrentWeight+item.Weight <= k.Capacity {
		k.Items = append(k.Items, item)
		k.CurrentWeight += item.Weight
		k.TotalValue += item.Value
		return true
	}
	return false
}

// GreedyKnapsackSolver solves the knapsack problem using a greedy heuristic.
type GreedyKnapsackSolver struct {
	Knapsack *Knapsack // The knapsack to store selected items.
	Items    []Item    // List of available items.
}

// NewGreedyKnapsackSolver initializes a new solver with the given capacity and items.
func NewGreedyKnapsackSolver(capacity int, items []Item) *GreedyKnapsackSolver {
	if len(items) == 0 {
		panic("Item list must not be empty.")
	}
	for i := range items {
		if items[i].Weight < 0 || items[i].Value < 0 {
			panic("Item values and weights must be non-negative.")
		}
		// Calculate value-to-weight ratio for sorting.
		if items[i].Weight == 0 {
			items[i].Ratio = 1e9 // Prioritize zero-weight items with a high ratio.
		} else {
			items[i].Ratio = float64(items[i].Value) / float64(items[i].Weight)
		}
	}
	return &GreedyKnapsackSolver{
		Knapsack: NewKnapsack(capacity),
		Items:    items,
	}
}

// Solve executes the greedy heuristic to fill the knapsack.
func (solver *GreedyKnapsackSolver) Solve() {
	// Sort items by value-to-weight ratio in descending order.
	sort.Slice(solver.Items, func(i, j int) bool {
		return solver.Items[i].Ratio > solver.Items[j].Ratio
	})

	// Add items to the knapsack while respecting capacity constraints.
	for _, item := range solver.Items {
		solver.Knapsack.AddItem(item)
	}
}

// Result provides the results of the heuristic solution.
func (solver *GreedyKnapsackSolver) Result() {
	fmt.Printf("Total Value: %d\n", solver.Knapsack.TotalValue)
	fmt.Printf("Total Weight: %d\n", solver.Knapsack.CurrentWeight)
	fmt.Println("Selected Items (Value, Weight):")
	for _, item := range solver.Knapsack.Items {
		fmt.Printf("  Value: %d, Weight: %d\n", item.Value, item.Weight)
	}
}

func main() {
	// Define the items (value, weight).
	items := []Item{
		{Value: 60, Weight: 10},
		{Value: 100, Weight: 20},
		{Value: 120, Weight: 30},
		{Value: 40, Weight: 0}, // Zero-weight item.
	}

	// Define the capacity of the knapsack.
	capacity := 50

	// Create a greedy knapsack solver.
	solver := NewGreedyKnapsackSolver(capacity, items)

	// Solve the problem.
	solver.Solve()

	// Display the results.
	solver.Result()
}

File: course_4/module_3/examples/greedy_knapsack_heuristic.py
class Item:
    """
    Represents an item in the knapsack problem.
    Each item has a value and a weight.
    """
    def __init__(self, value: int, weight: int):
        if value < 0 or weight < 0:
            raise ValueError("Value and weight must be non-negative integers.")
        self.value = value  # Value of the item.
        self.weight = weight  # Weight of the item.

    def value_per_weight(self) -> float:
        """
        Calculates and returns the value-to-weight ratio for the item.
        If the weight is zero, prioritize this item by returning a very high ratio.
        """
        if self.weight == 0:
            return float('inf')  # Zero-weight items are given priority.
        return self.value / self.weight


class Knapsack:
    """
    Represents the knapsack used to store selected items.
    Tracks the capacity of the knapsack and its current usage.
    """
    def __init__(self, capacity: int):
        if capacity <= 0:
            raise ValueError("Knapsack capacity must be a positive integer.")
        self.capacity = capacity  # Total capacity of the knapsack.
        self.items = []  # List of items included in the knapsack.

    def add_item(self, item: Item) -> bool:
        """
        Adds an item to the knapsack if there is enough remaining capacity.
        Returns True if the item was successfully added, False otherwise.
        """
        if self.current_weight() + item.weight <= self.capacity:
            self.items.append(item)  # Add the item to the knapsack.
            return True
        return False

    def current_weight(self) -> int:
        """
        Calculates the current total weight of the items in the knapsack.
        """
        return sum(item.weight for item in self.items)

    def total_value(self) -> int:
        """
        Calculates the total value of the items currently in the knapsack.
        """
        return sum(item.value for item in self.items)


class GreedyKnapsackSolver:
    """
    Solves the knapsack problem using a greedy heuristic.
    Selects items based on their value-to-weight ratio.
    """
    def __init__(self, capacity: int, items: list[Item]):
        if not items:
            raise ValueError("Item list must not be empty.")
        self.knapsack = Knapsack(capacity)  # Initialize the knapsack.
        self.items = items  # List of available items.

    def solve(self, criterion: str = "value_per_weight"):
        """
        Executes the greedy heuristic to solve the knapsack problem.
        Sorts items based on the specified criterion (default: value-to-weight ratio).
        """
        # Validate the sorting criterion.
        if criterion not in {"value_per_weight", "value", "weight"}:
            raise ValueError("Invalid sorting criterion. Must be 'value_per_weight', 'value', or 'weight'.")

        # Sort items by the specified criterion.
        if criterion == "value_per_weight":
            self.items.sort(key=lambda item: item.value_per_weight(), reverse=True)
        elif criterion == "value":
            self.items.sort(key=lambda item: item.value, reverse=True)
        elif criterion == "weight":
            self.items.sort(key=lambda item: item.weight)

        # Add items to the knapsack while respecting capacity constraints.
        for item in self.items:
            self.knapsack.add_item(item)

    def get_result(self) -> dict:
        """
        Returns the result of the heuristic as a dictionary.
        Includes the total value, the total weight, and the list of selected items.
        """
        return {
            "total_value": self.knapsack.total_value(),  # Total value of the selected items.
            "total_weight": self.knapsack.current_weight(),  # Total weight of the knapsack.
            "selected_items": [(item.value, item.weight) for item in self.knapsack.items],  # List of selected items.
        }


# Example usage:
if __name__ == "__main__":
    # Define the items (value, weight).
    items = [
        Item(value=60, weight=10),
        Item(value=100, weight=20),
        Item(value=120, weight=30),
        Item(value=40, weight=0),
    ]

    # Define the capacity of the knapsack.
    capacity = 50

    # Create a greedy knapsack solver.
    solver = GreedyKnapsackSolver(capacity, items)

    # Solve the problem.
    solver.solve()  # Use default sorting by value_per_weight.

    # Get and print the results.
    result = solver.get_result()
    print("Total Value:", result["total_value"])
    print("Total Weight:", result["total_weight"])
    print("Selected Items (Value, Weight):", result["selected_items"])
